pixels-core/src/main/java/io/pixelsdb/pixels/core/reader/PixelsRecordReaderBufferImpl.java

```diff
@@ -0,0 +1,360 @@
+     *
+     * @param size the number of rows in the row batch.
+     * @return the empty row batch.
+     */
+    private VectorizedRowBatch createEmptyEOFRowBatch(int size) {
+        TypeDescription resultSchema = TypeDescription.createSchema(new ArrayList<>());
+        VectorizedRowBatch resultRowBatch = resultSchema.createRowBatch(0, this.typeMode);
+        resultRowBatch.projectionSize = 0;
+        resultRowBatch.endOfFile = true;
+        resultRowBatch.size = size;
+        return resultRowBatch;
+    }
+
+    @Override
+    public VectorizedRowBatch readBatch() throws IOException {
+        long start = System.nanoTime();
+        if (!read()) {
+            return createEmptyEOFRowBatch(0);
+        }
+        readTimeNanos += System.nanoTime() - start;
+        curRowBatch = VectorizedRowBatch.deserialize(data);
+
+        LongColumnVector hiddenTimestampVector = (LongColumnVector) curRowBatch.cols[this.colNum - 1];
+        /**
+         * construct the selected rows bitmap, size is curBatchSize
+         * the i-th bit presents the curRowInRG + i row in chunkBuffers is selected or not.
+         */
+        int curBatchSize = curRowBatch.size;
+        Bitmap selectedRows = new Bitmap(curBatchSize, false);
+        int addedRows = 0;
+        for (int i = 0; i < curBatchSize; i++) {
+            if ((hiddenTimestampVector == null || hiddenTimestampVector.vector[i] <= this.option.getTransTimestamp())
+                    && (!checkBit(visibilityBitmap.get(fileIdIndex), i))) {
+                selectedRows.set(i);
+                addedRows++;
+            }
+        }
+        curRowBatch.applyFilter(selectedRows);
+        return curRowBatch;
+    }
+
+    @Override
+    public TypeDescription getResultSchema() {
+        // TODO(AntiO2): Schema evolution is currently not supported in Retina.
+        return typeDescription;
+    }
+
+    @Override
+    public boolean isValid() {
+        return false;
+    }
+
+    @Override
+    public boolean isEndOfFile() {
+        return endOfFile;
+    }
+
+    @Override
+    public boolean seekToRow(long rowIndex) throws IOException {
+        return false;
+    }
+
+    @Override
+    public boolean skip(long rowNum) throws IOException {
+        return false;
+    }
+
+    @Override
+    public long getCompletedRows() {
+        return 0;
+    }
+
+    @Override
+    public long getCompletedBytes() {
+        return 0;
+    }
+
+    @Override
+    public int getNumReadRequests() {
+        return fileIdIndex;
+    }
+
+    @Override
+    public long getReadTimeNanos() {
+        return readTimeNanos;
+    }
+
+    @Override
+    public long getMemoryUsage() {
+        return 0;
+    }
+
+    @Override
+    public void close() throws IOException {
+        scheduler.shutdown();
+    }
+
+    private String getMinioPathFromId(Integer id) {
+        return schemaName + '/' + tableName + '/' + id;
+    }
+
+    private void getMemtableDataFromMinio(String path) throws IOException {
```

Comment:
Rename this method to getMemtableDataFromStorage, as this class does not bind to the specific storage Minio.